### ISSUE TITLE
Minified file is very large

### ISSUE THREAD

**Issue by not-an-aardvark** [not-an-aardvark/snoowrap#20]
The [snoowrap.min.js](https://not-an-aardvark.github.io/snoowrap/snoowrap.min.js) file is 1.31 MB as of v0.11.6, which is extremely large for a minified file. There are a few reasons for this:

* snoowrap was originally created for Node, and while being able to run in the browser is a nice bonus, Browserify introduces a bit of boilerplate.
* Babel addes a bit of boilerplate as well.
* While snoowrap doesn't have many dependencies, a few of the dependencies that it does have (particularly [request-promise](https://www.npmjs.com/package/request-promise)) are very large.
* [UglifyJS](/mishoo/UglifyJS2), which is used to minify the file, has only preliminary support for ES6 syntax. Looking at the minified source, it seems like there are a bunch of variable names, etc. which aren't getting minified. It's possible that the minification will improve as development proceeds.

Summary of each component of the minified file (as of v0.11.6):

1. snoowrap (core): 55.719 KB (4.25%)
1. request-promise: 622.451 KB (47.5%)
1. bluebird: 64.007 KB (4.89%)
1. lodash: 46.127 KB (3.52%)
1. promise-chains: 22.923 KB (1.75%)
1. ws: 22.386 KB (1.71%)
1. other (browserify shims, etc.): 476.023 KB (36.3%)

The main contributor to the file size seems to be request-promise. It might be worth looking into whether request-promise can be split into smaller modules; it has a large number of features, and snoowrap is only actually using some of them. Another option would be to use a different request library.

**Comment by abdulhannanali** [comment]
Can't we just use `fetch` or simply `AJAXHttpRequest` on the client side instead of using request-promise?

**Comment by not-an-aardvark** [comment]
That's probably the best long-term solution, but there are a few issues with doing that right now:

* A few snoowrap functions, such as [`oauth_request`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#oauth_request__anchor), expose request-promise's API directly. In hindsight, this was probably a bad idea (since it means I can't change the request library that gets used internally), but getting rid of it would be a breaking change, so I won't be able to do that until I end up releasing 2.0.0.
* In terms of ease of development, it's easier to handle requests in a single way throughout the code. There isn't actually a separate codebase for the browser version of snoowrap -- the minified file is just generated from the node module by a one-line browserify script. As a result, implementing browser-only APIs separately would be a bit tedious. (It would also only be a partial solution, because if another module that depends on snoowrap gets converted with browserify, request-promise will end up in that bundle.)

I think the best solution for 2.0.0 would be to depend on node's `http` module directly, which should significantly decrease the bundle size while still working correctly with browserify.

**Comment by abdulhannanali** [comment]
IMHO, Using node's `http` module will make things more complex since request and request-promise modules are popular and well understood in the community. request also has a benefit of keeping things understandable for the developers who are coming to contribute. Can we separate the request logic for the client side and server side?

**Comment by not-an-aardvark** [comment]
That's another possibility. It will still make things a bit more complex, because I would probably want to shim at least a small subset of `request-promise`'s API to make the interface consistent. Also, while I could specifically exclude `request-promise` from the minified build, anyone who naively uses `browserify` on a package that has snoowrap as a dependency will end up including the `request-promise` module as well, since there will be at least one `require('request-promise')` statement in the code (even if it's in a conditional).

**Comment by not-an-aardvark** [comment]
Update: I ended up deprecating the use of `oauth_request` with most of its custom parameters, and I also deprecated the use of `snoowrap.min.js` in favor of versioned bundles such as `snoowrap-v1.6.0.min.js`. The versioned bundle does not contain `request-promise` (it has a small shim that uses XHR instead), so the minified version is only 300KB, which is much smaller than before. Since the versioned bundle didn't exist before this version, no one's site will be broken by the change.

The `snoowrap.js` and `snoowrap.min.js` files (without version labels) are deprecated and will remain at v1.5.0 indefinitely, to avoid breaking anything that used `oauth_request` previously. (It would have been necessary to eventually deprecate them anyway, since eventually upgrading them to 2.0 would have broken things.)
